### DISCUSSION TITLE
Confused about binding to axis swipes

### DISCUSSION THREAD

**Discussion by ndarilek** [Leafwing-Studios/leafwing-input-manager#218] [Q&A]
Hey, been struggling with this for a while now and can't seem to figure it out. I want to bind an action, `NavigationAction::SnapReverse`, to swiping down on the right stick. I have:

```
const DEADZONE: f32 = 0.5;
...
                    (
                        UserInput::Single(
                            SingleAxis {
                                axis_type: GamepadAxisType::RightStickY.into(),
                                positive_low: 2.,
                                negative_low: -DEADZONE,
                                value: None,
                            }
                            .into(),
                        ),
                        NavigationAction::SnapReverse,
                    ),
```

Unfortunately, this always triggers whenever I do anything on the right stick regardless of axis, and doesn't seem to respect the deadzone.

First, am I on the correct path with this binding?

If so, the syntax is a bit confusing. I don't care about the positive axis at all, but I still have to set an arbitrary value that my controller theoretically can't hit. Making these values `Option<_>` would be a bit clearer IMO.

Also, I'm not sure if these axis values are clamped, and am vaguely worried I'll get something out-of-bounds from either an off-brand gamepad sending wonky values, or some wild hardware that represents controls over a wider range. Not sure how realistic this concern is.

Thanks for any help.

**Comment by zicklag** [comment]
> Also, I'm not sure if these axis values are clamped, and am vaguely worried I'll get something out-of-bounds from either an off-brand gamepad sending wonky values, or some wild hardware that represents controls over a wider range. Not sure how realistic this concern is.

You can get the clamped axis values if you want by using [`clamped_axis_pair()`](https://docs.rs/leafwing-input-manager/latest/leafwing_input_manager/action_state/struct.ActionState.html#method.clamped_axis_pair). But be aware that if the gamepad outputs weird inputs, it will still trigger the axis on the weird inputs.

I think it's probably not _that_ big of a concern as long as you read the clamped output.

> First, am I on the correct path with this binding?

Absolutely. Here's what I do in my game for inputting the "down" direction on my menu screen:

```rust
            SingleAxis {
                axis_type: AxisType::Gamepad(GamepadAxisType::LeftStickY),
                positive_low: 1.0,
                negative_low: -0.5,
                value: None,
            }
```

( [link to code](https://github.com/zicklag/punchy/blob/a043a95ec04a26ca1f3f5ecfd031c9d77b5931db/src/loading.rs#L252-L260) )

Is your `DEADZONE` variable perhaps already a negative number, and the extra `-` is making the result positive? A positive number for `negative_low` would cause the axis to always trigger.

**Edit:** Nevermind, I just saw your DEADZONE const at the top of your codeblock. :thinking: Strange that it's not working, everything looks fine, I wonder if something else is wrong.

**Reply by ndarilek** [reply]
```
                    (
                        UserInput::Single(
                            SingleAxis {
                                axis_type: GamepadAxisType::RightStickY.into(),
                                positive_low: 1.,
                                negative_low: -1.,
                                value: None,
                            }
                            .into(),
                        ),
                        NavigationAction::SnapReverse,
                    ),
```

That triggers my action if I so much as tap the right stick in any direction.

I just updated the controller firmware and ran a test. It doesn't appear to be drifting.

**Reply by zicklag** [reply]
Strange, what happens if you try to navigate the menu of this web game with the left stick with your remote:

https://fishfolks.github.io/punchy/player/latest/

That's the game I'm using those bindings on ( but I am using a slightly older git version of the leafwing plugin ).

You should be able to select the button up and down by pressing up and down on the left stick.

**Reply by ndarilek** [reply]
Sorry, I'm blind and I suspect that game isn't accessible.

I'm going to try grabbing the axis value from within the system handling the action.

**Reply by zicklag** [reply]
I still have to debug, but I might be noticing some issues with axis inputs in my now, actually, it might be a regression in leafwing, but I have to experiment more. It seems that me pushing down on my joystick is moving the cursor up. I have to investigate more, it could still be a problem in my game, but I'll have to check.

**Comment by zicklag** [comment]
>  Making these values Option<_> would be a bit clearer IMO.

That was my initial thought, too, but other developers thought that it was more clear to avoid the option and go with a plain `f32`.

@alice-i-cecile Maybe we could open a poll about preference on that? I'm not sure if we could do that on GitHub discussions and maybe advertise it on Discord to try and get as many opinions as possible.

Not sure, just a thought.

**Reply by ndarilek** [reply]
It did give me pause as to what I should set the values I don't care about to. While I do know about `clamped_value`, I don't own the values this binding is getting, so I can't clamp them myself. Making values I don't care about `None` feels clearer than "I'm going to assign something arbitrary and desperately hope nothing ever accidentally goes over that value." :)
